pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,386 @@
+	} else if preEEName != curEEName {
+		if preEEName != "" {
+			err := c.deleteExternalEntity(preEn.Namespace, preEEName)
+			if err != nil {
+				return err
+			}
+		}
+		if curEEName != "" {
+			curEE := genExternalEntity(curEEName, curEn)
+			err := c.addExternalEntity(curEE)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		preIPs := sets.NewString(preEn.Spec.Interfaces[0].IPs...)
+		curIPs := sets.NewString(curEn.Spec.Interfaces[0].IPs...)
+		if (!reflect.DeepEqual(preEn.Labels, curEn.Labels)) || (!preIPs.Equal(curIPs)) {
+			eeName := genExternalEntityName(curEn)
+			updatedEE := genExternalEntity(eeName, curEn)
+			err := c.updateExternalEntity(updatedEE)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c.syncedExternalNode.Update(curEn)
+	return nil
+}
+func (c *ExternalNodeController) updateExternalEntity(ee *v1alpha2.ExternalEntity) error {
+	// resourceVersion must be specified for update operation,
+	// so it gets the existing ExternalEntity and modifies the changed fields.
+	existingEE, _ := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Get(context.TODO(), ee.Name, metav1.GetOptions{})
+	isChanged := false
+	if !reflect.DeepEqual(existingEE.Spec, ee.Spec) {
+		existingEE.Spec = ee.Spec
+		isChanged = true
+	}
+	if !reflect.DeepEqual(existingEE.Labels, ee.Labels) {
+		existingEE.Labels = ee.Labels
+		isChanged = true
+	}
+	if isChanged {
+		_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Update(context.TODO(), existingEE, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(namespace string, name string) error {
+	obj, exists, _ := c.syncedExternalNode.GetByKey(k8s.NamespacedName(namespace, name))
+	if !exists {
+		klog.InfoS("Skipping deleting ExternalNode as it does not exist", "enName", namespace, "enNamespace", namespace)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	eeName := genExternalEntityName(en)
+	if eeName == "" {
+		klog.InfoS("Interfaces are empty for ExternalNode", "ExternalNode", klog.KObj(en))
+		return nil
+	}
+	err := c.deleteExternalEntity(namespace, eeName)
+	if err != nil {
+		return err
+	}
+	c.syncedExternalNode.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalEntity(namespace string, name string) error {
+	err := c.crdClient.CrdV1alpha2().ExternalEntities(namespace).Delete(context.TODO(), name, metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.InfoS("Skipping deleting ExternalEntity as it is not found", "eeName", name, "eeNamespace", namespace)
+		return nil
+	}
+	return err
+}
+
+func genExternalEntityName(externalNode *v1alpha1.ExternalNode) string {
+	if len(externalNode.Spec.Interfaces) == 0 {
+		return ""
+	}
+	// Only one network interface is supported now.
+	// Other interfaces except interfaces[0] will be ignored if there are more than one interfaces.
+	ifName := externalNode.Spec.Interfaces[0].Name
+	if ifName == "" {
```

Comment:
Could this happen?

Reply:
Yes, we allow users only specify IP without name. Actually based on the discussion with Wenying on previous PR https://github.com/antrea-io/antrea/pull/3687, IP is the must have field for ExternalNode case.

Reply:
I also move this function to util/externalnode as the mapping will be used for agent side as well.

Reply:
Ok. Got it.